internal/driver/custommetadata.go

```diff
@@ -0,0 +1,124 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package driver
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+)
+
+func (onvifClient *OnvifClient) initCustomMetadata(device contract.Device) contract.Device {
+	if _, found := device.Protocols[CustomMetadata]; !found {
+		saveOnvifProtocol := device.Protocols[OnvifProtocol]
+		device.Protocols = map[string]contract.ProtocolProperties{
+			OnvifProtocol:  saveOnvifProtocol,
+			CustomMetadata: {},
+		}
+	}
+	return device
+}
```

Comment:
Can be simplified, and also I think may make more sense to be a pointer:

```suggestion
func (onvifClient *OnvifClient) initCustomMetadata(device *contract.Device) {
	if _, found := device.Protocols[CustomMetadata]; !found {
		device.Protocols[CustomMetadata] = contract.ProtocolProperties{}
	}
}
```

Reply:
I can't assign anything to device.Protocols[CustomMetadata] in the case that CustomMetadata doesn't exist, because it will cause a panic. It's almost guaranteed to exist at this point unless the user has predefined devices that don't include CustomMetadata, as this is already initialized in discovery. That's why I was resetting the entire device.Protocols. Do you see an alternative to what I had in that case?

Reply:
You can't set a value on the `CustomMetadata`, but you can set `CustomMetadata` itself. See the difference:

```go
device.Protocols[CustomMetadata]["Location"] = "Exit"   <== BAD

device.Protocols[CustomMetadata] = contract.ProtocolProperties{}   <== OK

device.Protocols[CustomMetadata] = contract.ProtocolProperties{
    "Location": "Exit",
}   <== OK

```

Reply:
I see my issue now, the panic I was seeing was due to something in my test setup. Thank you!